### PR TITLE
kvclient: change the default circuit breaker to liveness

### PR DESCRIPTION
**Pull request by andrewbaptist**
We previously disabled circuit breakers in 24.1 due to a lack of scale testing and added a nob for liveness only. This PR changes the default to liveness and adds a test to validate that it works successfully.

Epic: none
Informs: #121941

Release note: None

### CODE REVIEW

**Comment by andrewbaptist** [comment]
@kvoli We can chat on this PR on Monday if you want me to walk you through it.

File: pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go [hunk 1]
```diff
@@ -121,6 +123,143 @@ func TestDistSenderReplicaStall(t *testing.T) {
...
+						// If the circuit breaker on the liveness range is disabled,
+						// we don't expect to ever see timeout or other errors as it
+						// should retry in DistSender and not get stuck.
+						require.False(t, scanError)
+						// Also verify liveness updates propagate by gossip. This
+						// validates that a new leader is established and broadcasts the
+						// liveness information.
+						testutils.SucceedsSoon(t, func() error {
+							t.Logf("read liveness from gossip on n2")
+							node0Cache := tc.Server(1).NodeLiveness().(*liveness.NodeLiveness).GetNodeVitalityFromCache(1)
```
Comment (andrewbaptist) [inline pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go:251]: Update this to node2Cache.

**Review by kvoli — COMMENTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123811)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmP9un2SrptQSovLC-):*
nit: knob
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmiKFU0pGbaJn-y2tJ:-NxmiKFU0pGbaJn-y2tK:b-355i1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L126)):*
> ```Go
> }
> 
> // TestDistSenderCircuitBreakerModes validates the behavior ot DistSender after
> ```

nit: of
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmiQkF3E0By0fE0zdm:-NxmiQkF3E0By0fE0zdn:bg9sylo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L128)):*
> ```Go
> // TestDistSenderCircuitBreakerModes validates the behavior ot DistSender after
> // a range stall caused by a locked mutex. It tests two different ranges
> // (liveness and regular) and the three different circuit breaker modes.
> ```

nit: either enumerate the circuit breaker modes or mention its all of them.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-Nxmizlp05ZGH8FuW3-4:-Nxmizlp05ZGH8FuW3-5:bkw02xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L174)):*
> ```Go
> 				desc := tc.AddVotersOrFatal(t, keys.NodeLivenessPrefix, tc.Targets(1, 2)...)
> 				tc.TransferRangeLeaseOrFatal(t, desc, tc.Target(2))
> 				// Ensure the n1 dist sender cache is up-to-date.
> ```

It would be helpful to explain why the distsender cache needs to be up to date.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-Nxmip2y7xX2P308NR2j:-Nxmip2y7xX2P308NR2k:b-64tler) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L179)):*
> ```Go
> 				// Upreplicate the liveness range and put on n3.
> 				if scratchRange {
> 					// We only use the scratch range for the non-livenes mode.
> ```

nit: non-liveness
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjCWw1o2LzOwg_rEy:-NxmjCWw1o2LzOwg_rEz:b-8ilqey) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L190)):*
> ```Go
> 					// Ensure the n1 dist sender cache is up-to-date.
> 					tc.WaitForValues(t, key, []int64{1, 1, 1})
> 					t.Logf("moved lease to n3")
> ```

nit: this can just be `t.Log(...)`.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjFxVAm3YvoiD3pT3:-NxmjFxVAm3YvoiD3pT4:b-q9jcdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L199)):*
> ```Go
> 				mu.Lock()
> 				defer mu.Unlock()
> 				t.Logf("deadlocked range on n3")
> ```

Same as above, it might be helpful to include the replica descriptor if sticking with `t.Logf`.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjPl96_-17hYUoyyK:-NxmjPl96_-17hYUoyyL:bwd2atw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L203)):*
> ```Go
> 				if scratchRange {
> 					// Perform a read from n1, which will stall. Eventually, the lease will be
> 					// picked up by a different node. The DistSender should detect this and retry
> ```

nit: `dist sender` is used elsewhere in this test but `DistSender` is used here.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 205 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjbEV8ASr8SN4vhjb:-NxmjbEV8ASr8SN4vhjc:b-vl2cqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L205)):*
> ```Go
> 					// picked up by a different node. The DistSender should detect this and retry
> 					// the read there.
> 					t.Logf("sending Get request to blocked range")
> ```

nit: `t.Log(...)`
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjjjQCJ8ySNrSQTio:-NxmjjjQCJ8ySNrSQTip:b-x94dr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L214)):*
> ```Go
> 						require.Error(t, err)
> 					}
> 				} else {
> ```

Consider adding a comment just below here stating we are hitting the liveness range.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 218 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-Nxmk4WcBjousOg-v62h:-Nxmk4WcBjousOg-v62i:b-ovdfi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L218)):*
> ```Go
> 					// Check that we get the correct value directly from KV as well as
> 					// from gossip. The leases may end up on either n1 or n2.
> 					testutils.SucceedsSoon(t, func() error {
> ```

Add a comment describing why the succeeds soon is necessary.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmkbQG6ZTWZXkqexd-:-NxmkbQG6ZTWZXkqexd0:bb4ldus) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L242)):*
> ```Go
> 						require.True(t, scanError)
> 					} else {
> 						// If the circuit breaker on the liveness range is disabled,
> ```

Should this be enabled, not disabled?

**Review by andrewbaptist — COMMENTED** [review]
TFTR!

Updated all the comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-:-NxsDXPOCEJjYWlEPd-W:bx7rkcb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmP9un2SrptQSovLC-):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: knob
</blockquote></details>

Fixed
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmiKFU0pGbaJn-y2tJ:-NxsBOUE30J73p4GKhrq:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L126)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: of
</blockquote></details>

fixed
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmiQkF3E0By0fE0zdm:-NxsBY-O53JJQvc_Pqar:b4o2uqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L128)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: either enumerate the circuit breaker modes or mention its all of them.
</blockquote></details>

Added that its all of them.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-Nxmizlp05ZGH8FuW3-4:-NxsBp6y-g8tSapIVJgN:b-8b11j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L174)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

It would be helpful to explain why the distsender cache needs to be up to date.
</blockquote></details>

Added
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-Nxmip2y7xX2P308NR2j:-NxsBrcT9p9pMm8QXqVl:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L179)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: non-liveness
</blockquote></details>

Fixed
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjCWw1o2LzOwg_rEy:-NxsBw_u5GSXZkjiq-Ca:bie8mlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L190)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: this can just be `t.Log(...)`.
</blockquote></details>

Fixed (and the other ones).
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjFxVAm3YvoiD3pT3:-NxsCAanBGFbwY8PAWpK:b8qc1vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L199)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Same as above, it might be helpful to include the replica descriptor if sticking with `t.Logf`.
</blockquote></details>

Agreed - I added the descriptor and left it as Logf. I also fixed all the Log/Logf inconsistencies in this file.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjPl96_-17hYUoyyK:-NxsCIv61ALxvqVkzAxm:b1yn53f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L203)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: `dist sender` is used elsewhere in this test but `DistSender` is used here.
</blockquote></details>

I changed it to DistSender everywhere. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 205 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjbEV8ASr8SN4vhjb:-NxsC_Fq9bmMMlU51QWE:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L205)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: `t.Log(...)`
</blockquote></details>

Done
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmjjjQCJ8ySNrSQTio:-NxsCf-Q5Ls45S5rbcrU:b-8b11j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L214)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Consider adding a comment just below here stating we are hitting the liveness range.
</blockquote></details>

Added
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 218 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-Nxmk4WcBjousOg-v62h:-NxsCfU58bce_Dk5LoY7:b-kuu4nm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L218)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Add a comment describing why the succeeds soon is necessary.
</blockquote></details>

Good call added. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-NxmkbQG6ZTWZXkqexd-:-NxsDF956qleWyVbdpsX:bqooe1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L242)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Should this be enabled, not disabled?
</blockquote></details>

yes - enabled, fixed.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go` line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123811#gh-1598623561:-NxsDZNrDNpPdYs8mEon:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e59dbf82482d36befd9dc593bcfd244221362c8/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go#L251)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Update this to node2Cache.
</blockquote></details>

Changed.

**Review by kvoli — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123811#-:-NxsMbD7BpMy0QhbZSfz:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained

**Pull request merged**